### ISSUE TITLE
NPE in CodeStream.retrieveLocalType

### ISSUE THREAD

**Issue by cdietrich** [eclipse-jdt/eclipse.jdt.core#2428]
with the newest Xtext nightly we see the following error in one of our tests

```
Caused by: java.lang.NullPointerException: Cannot read the array length because the return value of "org.eclipse.jdt.internal.compiler.lookup.ReferenceBinding.syntheticOuterLocalVariables()" is null
	at org.eclipse.jdt.internal.compiler.codegen.CodeStream.retrieveLocalType(CodeStream.java:7686)
	at org.eclipse.jdt.internal.compiler.codegen.CodeStream.pushTypeBinding(CodeStream.java:7705)
	at org.eclipse.jdt.internal.compiler.codegen.CodeStream.aload_3(CodeStream.java:394)
	at org.eclipse.jdt.internal.compiler.codegen.CodeStream.load(CodeStream.java:6284)
	at org.eclipse.jdt.internal.compiler.codegen.CodeStream.load(CodeStream.java:6186)
	at org.eclipse.jdt.internal.compiler.ast.SingleNameReference.generateCode(SingleNameReference.java:520)
	at org.eclipse.jdt.internal.compiler.ast.Statement.generateArguments(Statement.java:440)
	at org.eclipse.jdt.internal.compiler.ast.ExplicitConstructorCall.generateCode(ExplicitConstructorCall.java:177)
	at org.eclipse.jdt.internal.compiler.ast.ConstructorDeclaration.internalGenerateCode(ConstructorDeclaration.java:441)
	at org.eclipse.jdt.internal.compiler.ast.ConstructorDeclaration.generateCode(ConstructorDeclaration.java:312)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:762)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:806)
	at org.eclipse.jdt.internal.compiler.ast.QualifiedAllocationExpression.generateCode(QualifiedAllocationExpression.java:248)
	at org.eclipse.jdt.internal.compiler.ast.MessageSend.generateCode(MessageSend.java:618)
	at org.eclipse.jdt.internal.compiler.ast.Statement.generateArguments(Statement.java:440)
	at org.eclipse.jdt.internal.compiler.ast.MessageSend.generateCode(MessageSend.java:625)
	at org.eclipse.jdt.internal.compiler.ast.ReturnStatement.generateCode(ReturnStatement.java:234)
	at org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration.generateCode(AbstractMethodDeclaration.java:388)
	at org.eclipse.jdt.internal.compiler.ast.AbstractMethodDeclaration.generateCode(AbstractMethodDeclaration.java:324)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:762)
	at org.eclipse.jdt.internal.compiler.ast.TypeDeclaration.generateCode(TypeDeclaration.java:832)
	at org.eclipse.jdt.internal.compiler.ast.CompilationUnitDeclaration.generateCode(CompilationUnitDeclaration.java:412)
	at org.eclipse.jdt.internal.compiler.Compiler.process(Compiler.java:915)
	at org.eclipse.jdt.internal.compiler.Compiler.processCompiledUnits(Compiler.java:574)
	at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:474)
	at org.eclipse.jdt.internal.compiler.Compiler.compile(Compiler.java:425)
	at org.eclipse.xtext.xbase.testing.InMemoryJavaCompiler.compile(InMemoryJavaCompiler.java:280)
	at org.eclipse.xtend.core.tests.compiler.CompilerTestHelper.compileToClass(CompilerTestHelper.java:122)
```

the code compiled is

```
package foo;

import testdata.ClassWithVarArgs;

@SuppressWarnings("all")
public class Test {
  public Object foo() throws Exception {
    return Boolean.valueOf(new ClassWithVarArgs(1, "", "") {
    }.isVarArgConstructor());
  }
}
```

with

```
/*******************************************************************************
 * Copyright (c) 2011, 2016 itemis AG (http://www.itemis.eu) and others.
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0.
 *
 * SPDX-License-Identifier: EPL-2.0
 *******************************************************************************/
package testdata;

import java.util.ArrayList;
import java.util.List;

import com.google.common.collect.Lists;

/**
 * @author Sebastian Zarnekow - Initial contribution and API
 */
public class ClassWithVarArgs {

	private boolean varArgConstructor;
	private boolean defaultConstructor;

	public ClassWithVarArgs() {
		defaultConstructor = true;
	}
	
	public ClassWithVarArgs(int i, String... strings) {
		this.varArgConstructor = true;
	}
	
	public ClassWithVarArgs(int i, String s1, String s2) {
		this.varArgConstructor = false;
	}
	
	public List<String> stringsToList(String... strings) {
		if (strings == null)
			return null;
		return Lists.newArrayList(strings);
	}
	
	public List<String> stringsToList2(String firstArg, String... strings) {
		if (strings == null)
			return null;
		final ArrayList<String> result = Lists.newArrayList(firstArg);
		result.addAll(Lists.newArrayList(strings));
		return result;
	}
	
	@SuppressWarnings("unchecked")
	public <T> List<T> toList(T... args) {
		return Lists.newArrayList(args);
	}
	
	@SuppressWarnings("unchecked")
	public <T extends Number> List<T> toNumberList(T... args) {
		return Lists.newArrayList(args);
	}
	
	public List<String> stringsToList(String s1, String s2) {
		return Lists.newArrayList("foo", s1, s2);
	}
	
	public boolean isVarArgConstructor() {
		return varArgConstructor;
	}
	
	public boolean isDefaultConstructor() {
		return defaultConstructor;
	}
	
	public String logInfo(String msg) {
		return String.format("logInfo(%s)", msg);
	}

	public String logInfo(String msg, Object... args) {
		return String.format("logInfo(%s, args...)", msg);
	}
	
	public String logInfo2(String msg) {
		return String.format("logInfo2(%s)", msg);
	}

	public String logInfo2(String msg, String param1) {
		return String.format("logInfo2(%s, %s)", msg, param1);
	}
	
	public String logInfo2(String msg, String param1, String param2) {
		return String.format("logInfo2(%s, %s, %s)", msg, param1, param2);
	}
	
	public String logInfo2(String msg, Object... args) {
		return String.format("logInfo2(%s, args...)", msg);
	}
	
	private List<String> values;
	
	public void setValues(String ... values) {
		this.values = stringsToList(values);
	}
	
	public List<String> getValues() {
		return values;
	}
	
}

```

unfortunately i cannot reproduce this yet with a java main

**Comment by cdietrich** [comment]
hmmm i just checked out jdt master as source instead of i-builds and with that it works.
so maybe the next ibuild will fix it

**Comment by srikanth-sankaran** [comment]
At what language level is your code compiled ? There was a fix for a problem with compiling code at < 1.6 that was made yesterday.

**Comment by cdietrich** [comment]
we still use java 5 as default

**Comment by LorenzoBettini** [comment]
> At what language level is your code compiled ? There was a fix for a problem with compiling code at < 1.6 that was made yesterday.

Do you know whether https://download.eclipse.org/eclipse/updates/4.32-I-builds/I20240508-1800/plugins/ already contains the fix?

**Comment by srikanth-sankaran** [comment]
> unfortunately i cannot reproduce this yet with a java main

I don't know what you mean by java main - you have given enough to reproduce the problem. Apart from compiling the given code at levels < 1.6, you also need to turn off the option `Add variable attributes to generated class files` in the project properties dialog under `Java Compiler`

I can confirm this is the same problem fixed by commit id 6bd9ca2b9f8b2aadc0ce8ec99367ca2fc6a7d396 pushed yesterday. I.e by doing "time travel" I am able to reproduce the problem before commit 6bd9ca2b9f8b2aadc0ce8ec99367ca2fc6a7d396 and confirm the problem is fixed after 6bd9ca2b9f8b2aadc0ce8ec99367ca2fc6a7d396 

So master HEAD has the fix already
